### PR TITLE
ARROW-2417: [Rust] Fix API safety issues

### PR DESCRIPTION
**Pull request by andygrove**
I reviewed all uses of unsafe in the API implementation and added appropriate assertions where needed to guarantee that the API we expose is safe. Also added tests to verify in some cases.

### CODE REVIEW

File: rust/src/buffer.rs [hunk 1]
```diff
@@ -32,6 +32,7 @@ pub struct Buffer<T> {
 }
 
 impl<T> Buffer<T> {
+    /// create a buffer from an existing region of memory (must already be byte-aligned)
     pub fn from_raw_parts(data: *const T, len: i32) -> Self {
```
Comment (jesskfullwood) [inline rust/src/buffer.rs]: I believe this should be marked unsafe - if an incorrect pointer or len is passed, `get` and `slice` could access arbitrary memory.

File: rust/src/buffer.rs [hunk 2]
```diff
@@ -41,24 +42,27 @@ impl<T> Buffer<T> {
...
 
+    /// Get a pointer to the data contained by the buffer
     pub fn data(&self) -> *const T {
         self.data
     }
 
     pub fn slice(&self, start: usize, end: usize) -> &[T] {
         assert!(start <= end);
-        assert!(start <= self.len as usize);
+        assert!(start < self.len as usize);
```
Comment (jesskfullwood) [inline rust/src/buffer.rs]: I do not think this assert is necessary - if start == end then the slice has length zero so is safe.

File: rust/src/buffer.rs [hunk 3]
```diff
@@ -41,24 +42,27 @@ impl<T> Buffer<T> {
...
+        assert!(i < self.len as usize);
         unsafe { &(*self.data.offset(i as isize)) }
     }
 
-    /// Deprecated method (used by Bitmap)
+    /// Write to a slot in the buffer
     pub fn set(&mut self, i: usize, v: T) {
+        assert!(i < self.len as usize);
         unsafe {
             let p = mem::transmute::<*const T, *mut T>(self.data);
```
Comment (jesskfullwood) [inline rust/src/buffer.rs]: `transmute` is overkill here - can get away with (safe) `let p = self.data as *mut _`

File: rust/src/buffer.rs [hunk 4]
```diff
@@ -41,24 +42,27 @@ impl<T> Buffer<T> {
...
         assert!(start <= end);
-        assert!(start <= self.len as usize);
+        assert!(start < self.len as usize);
         assert!(end <= self.len as usize);
         unsafe { slice::from_raw_parts(self.data.offset(start as isize), (end - start) as usize) }
     }
 
     /// Get a reference to the value at the specified offset
     pub fn get(&self, i: usize) -> &T {
+        assert!(i < self.len as usize);
```
Comment (jesskfullwood) [inline rust/src/buffer.rs]: panic message would be useful - `assert!(..., "Out-of-bounds access")` (same goes for elsewhere)

File: rust/src/buffer.rs [hunk 5]
```diff
@@ -41,24 +42,27 @@ impl<T> Buffer<T> {
...
     pub fn slice(&self, start: usize, end: usize) -> &[T] {
         assert!(start <= end);
-        assert!(start <= self.len as usize);
+        assert!(start < self.len as usize);
         assert!(end <= self.len as usize);
         unsafe { slice::from_raw_parts(self.data.offset(start as isize), (end - start) as usize) }
     }
 
     /// Get a reference to the value at the specified offset
     pub fn get(&self, i: usize) -> &T {
```
Comment (jesskfullwood) [inline rust/src/buffer.rs]: Good idea to combine `slice` and `get`, to mirror the stdlib [slice api](https://doc.rust-lang.org/std/primitive.slice.html#method.get) (it returns a slice if passed a `Range`, item if passed a `usize`. Depending on how `Buffer` is used, might also be worth implementing `Index` and `IndexMut`

File: rust/src/buffer.rs [hunk 6]
```diff
@@ -159,7 +166,7 @@ impl From<Bytes> for Buffer<u8> {
                     mem::transmute::<*const u8, *const libc::c_void>(bytes.as_ptr()),
                     len * sz,
                 );
-                mem::transmute::<*mut libc::c_void, *const u8>(dst)
+                mem::transmute::<*mut libc::c_void, *mut u8>(dst)
```
Comment (jesskfullwood) [inline rust/src/buffer.rs]: Again, `transmute` is overkill, a cast would do (check out the [nomicon](https://doc.rust-lang.org/nightly/nomicon/casts.html).

**Comment by jesskfullwood** [comment]
Made a few general comments

**Comment by andygrove** [comment]
Thanks @jesskfullwood I have pushed changes to address the feedback. It's nice to get rid of those transmute calls!

**Comment by xhochy** [comment]
@jesskfullwood Does this look good to you?

**Comment by jesskfullwood** [comment]
Yes, LGTM!

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**